# Highly Specific Gene Regulatory Networks Determine Quantitative Disease Resistance of Wild Tomatoes Against a Necrotrophic Pathogen

## Authors

**Severin Einspanier**, Department of Phytopathology and Crop Protection, Institute of Phytopathology, Faculty of Agricultural and Nutritional Sciences, Christian Albrechts University, Kiel, Germany  
**Christopher Tominello-Ramirez**, Department of Phytopathology and Crop Protection, Institute of Phytopathology, Faculty of Agricultural and Nutritional Sciences, Christian Albrechts University, Kiel, Germany  
**Florent Delplace**, Laboratoire des Interactions Plantes Microorganismes Environnement (LIPME), INRAE, CNRS, Castanet Tolosan Cedex, France  
**Remco Stam**, Department of Phytopathology and Crop Protection, Institute of Phytopathology, Faculty of Agricultural and Nutritional Sciences, Christian Albrechts University, Kiel, Germany

Please address correspondence to: s.einspanier@phytomed.uni-kiel.de or r.stam@phytomed.uni-kiel.de

## ReadMe 

This is the code repository for all scripts involved in the study. We used 3' UTR sequencing technology to characterize the regulatory response to *Sclerotinia sclerotiorum* infection and to extrapolate regulatory cues influencing the level of quantitative disease resistance. 

This repository is structured following the tools/methods used:
1. Bioinformatic pipeline
2. Basic DESeq analysis
3. GenEra (phylostratigraphic analysis)
4. Orthogroup analysis
5. GO-terms
6. Weighted gene correlation network analysis (WGCNA)
7. Gene regulatory networks (GRN)
8. Phylotranscriptomics (TAI/TDI)

## Data Availability 

